rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,410 @@
+10 minutes), and stored in the backend. If the database engine supports it, the
+sync may also happen when a schema change is detected. For example, in Postgres
+we case use
+(trigger+notify)[https://medium.com/launchpad-lab/postgres-triggers-with-listen-notify-565b44ccd782].
+
+#### Import result: the `db_object` resource
+
+Import process creates a number of `db_object` resources with applied labels.
+Aside from standard metadata fields, the object spec consists of a number of
+predefined attributes:
+
+```protobuf
+// DatabaseObjectSpec is the spec for the database object.
+message DatabaseObjectSpec {
+  string Protocol = 1 [(gogoproto.jsontag) = "protocol,omitempty"];
+  string DatabaseServiceName = 2 [(gogoproto.jsontag) = "db_service_name,omitempty"];
+  string ObjectKind = 3 [(gogoproto.jsontag) = "object_kind,omitempty"];
+  string Database = 4 [(gogoproto.jsontag) = "database,omitempty"];
+  string Schema = 5 [(gogoproto.jsontag) = "schema,omitempty"];
+  string Name = 6 [(gogoproto.jsontag) = "name,omitempty"];
+}
+```
+
+All of the above fields are optional, and more may be added in the future if
+needed. The database-specific implementation is responsible for populating these
+fields from the database schema.
+
+### Database object permissions
+
+The permissions for particular objects are defined in a role using the new
+`db_permissions` field, found under `spec.allow` and `spec.deny` respectively.
+Each permission specifies:
+
+- A list of labels the database object must match.
+- A list of permissions that will be given to the object for the user.
+
+As an example, the role `db_read_non_confidential` allows access to tables and
+views with a label `confidential: false` and explicitly disallows any access to
+those labeled `confidential: true`.
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: db_read_non_confidential
+spec:
+  allow:
+    db_permissions:
+      - match:
+          confidential: 'false'
+          kind:
+            - table
+            - view
+        permissions:
+          - SELECT
+          - INSERT
+          - UPDATE
+  deny:
+    db_permissions:
+      - match:
+          confidential: 'true'
+        permissions:
+          - '*'
+```
+
+The object attributes are not replicated automatically as labels. In the example
+above, `kind` is a label which must be applied by an import rule, for example
+one as follows:
+
+```yaml
+kind: db_object_import_rule
+version: v1
+metadata:
+  name: import_object_kind
+spec:
+  mappings:
+    - add_labels:
+        kind: table
+      match:
+        object_kinds:
+          - table
+    - add_labels:
+        kind: view
+      match:
+        object_kinds:
+          - view
+    - add_labels:
+        kind: schema
+      match:
+        object_kinds:
+          - schema
+  priority: 100
+```
+
+#### Applying permissions
+
+The permissions will be applied to the user after the user is provisioned in the
```

Comment:
I think we're missing explanation of _how_ these permissions will actually be applied to the created user. Will we run appropriate "grant" statements for the user?

Reply:
For Postgres or MySQL, yes. But the exact mechanism is database dependent by its very nature. I'll clarify this using Postgres as an example, though.

Reply:
I've added some text explaining this.